Define operator Doc (#130)
operator-framework/olm-docs: content/en/docs/Reference/operator.md

---
title: "Operators"
linkTitle: "Operators"
weight: 3

> date: 2021-03-15

Reviewer: I think it is missing the description of the doc as well.

---

> <br>

Reviewer 2: You need to remove this inline html as it's breaking the markdown rendering of the next header.

> ## What is an Operator?

Reviewer 2: Super minor nit, can we add a newline between the ending title and this header?

Reviewer 3: done

> Operators make it easy to manage complex stateful applications on top of Kubernetes. By [Extendig the Kubernetes API with CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) it is possible to create what we call Operator Projects where you will be able to create and define your own API's to represent your solutions on the cluster and managed them by looking for the desired state. For a better understanding see [Operator Pattern][operator-pattern]

Reviewer 2: ```suggestion
Operators make it easy to manage complex stateful applications on top of Kubernetes. By [extending the Kubernetes API with CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) it is possible to create what we call Operator Projects where you will be able to create and define your own API's to represent your solutions on the cluster and manage them by looking for the desired state. For a better understanding see the [Operator Pattern][operator-pattern].
```

Reviewer 2: Do we also want to code bracket, `CustomResourceDefinitions`? WDYT?

> Operators make it easy to manage complex stateful applications on top of Kubernetes. By [Extending the Kubernetes API](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) with CustomResourceDefinitions it is possible to create what we call Operator Projects where you will be able to create and define your own API's to represent your solutions on the cluster and managed them by looking for the desired state. For a better understanding see [Operator Pattern][operator-pattern]

Reviewer 2: ```suggestion
Operators make it easy to manage complex stateful applications on top of Kubernetes. By [Extending the Kubernetes API](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) with CustomResourceDefinitions it is possible to create what we call Operator Projects where you will be able to create and define your own API's to represent your solutions on the cluster and managed them by looking for the desired state. For a better understanding, see the [Operator Pattern][operator-pattern].
```

## What is the difference between a Controller and an Operator

> See that following the [Operator's pattern][operator-pattern]; you will create [Controllers](https://kubernetes.io/docs/concepts/architecture/controller/) which provides a reconcile function responsible for synchronizing the resources until a desired state on the cluster. For a more practical explanation, see also [What’s in a controller?](https://book.kubebuilder.io/cronjob-tutorial/controller-overview.html#whats-in-a-controller)

Reviewer 2: ```suggestion
See that following the [Operator's pattern][operator-pattern]; you will create [Controllers](https://kubernetes.io/docs/concepts/architecture/controller/) which provides a reconcile function responsible for synchronizing the resources until a desired state on the cluster. For a more practical explanation, see also [What’s in a controller?](https://book.kubebuilder.io/cronjob-tutorial/controller-overview.html#whats-in-a-controller).
```

## What is Operator Lifecycle Manager?

Operator Lifecycle Manager (OLM) helps users install, update, and manage the lifecycle of all Operators and their associated services running across their clusters. This project is a component of the [Operator Framework][operator-framework], an open-source toolkit to manage Kubernetes native applications, called Operators, in an effective, automated, and scalable way. Read more in the introduction [blog post](https://www.openshift.com/blog/introducing-the-operator-framework?extIdCarryOver=true&sc_cid=701f2000001Css5AAC).

> ## How does OLM enable cluster admins and Developers?

Reviewer 2: Super minor nit, can we normalize the capitalization of non-articles in this head description?

```suggestion
## How does OLM enable cluster admins and developers?
```

Or,

```suggestion
## How Does OLM Enable Cluster Admins and Developers?
```

Reviewer 3: I have made all the changes that you have suggested. Thanks

Reviewer 3: Thanks for all the suggestions. I have pushed all the changes.

> The OLM is enabled by default in OpenShift Container Platform 4.X, which aids cluster administrators in installing, upgrading, and granting access to Operators running on their cluster. The OpenShift Container Platform web console provides management screens for cluster administrators to install Operators, as well as grant specific projects access to use the catalog of Operators available on the cluster.

Reviewer 2: ```suggestion
The OLM is enabled by default in OpenShift Container Platform 4.x, which aids cluster administrators in installing, upgrading, and granting access to Operators running on their cluster. The OpenShift Container Platform (OCP) web console provides management screens for cluster administrators to install Operators, as well as grant specific projects access to use the catalog of Operators available on the cluster.
```

Reviewer 2: Hmm, we may want to just document the full name, like Operator Lifecycle Manager, or OpenShift Container Platform once, and then abbreviate the rest of those references throughout this document. WDYT?

Reviewer 3: I have checked for both. there arent many repetitions and also some of them are in the heading

Reviewer 2: Cool, just wanted to double-check. Thanks :+1:.

For developers, a self-service experience allows provisioning and configuring instances of databases, monitoring, and big data services without having to be subject matter experts, because the Operator has that knowledge baked into it.

> 
> 

Reviewer 2: ```suggestion

```

## Build with the Operator SDK

[Operator SDK][osdk] is also a component of the [Operator Framework][operator-framework]. Writing an operator today can be difficult because of challenges such as using low-level APIs, writing boilerplate, and a lack of modularity which leads to duplication.

> The [Operator SDK][osdk] is a framework that uses the [controller-runtime][controller-runtime] library to make writing operators easier by providing:
> - High-level APIs and abstractions to write the operational logic more intuitively
> - Tools for scaffolding and code generation to bootstrap a new project fast
> - Extensions to cover common operator use cases

Reviewer 2: Looks like the linting check is complaining about needing a newline between body content and the first list element.

```suggestion
The [Operator SDK][osdk] is a framework that uses the [controller-runtime][controller-runtime] library to make writing operators easier by providing:

- High-level APIs and abstractions to write the operational logic more intuitively
- Tools for scaffolding and code generation to bootstrap a new project fast
- Extensions to cover common operator use cases
```

> Note that [Operator SDK][osdk] provides helpers and features to help you [integrate your project with OLM][olm-integration].  For further information check https://sdk.operatorframework.io/ .

Reviewer 2: ```suggestion
Note that [Operator SDK][osdk] provides helpers and features to help you [integrate your project with OLM][olm-integration].
```

> 
> 

Reviewer 2: ```suggestion

```

## Package with the Operator Lifecycle Manager

With OLM, administrators can control which Operators are available in what namespaces and who can interact with running Operators. The permissions of an Operator are accurately configured automatically to follow a least-privilege approach. OLM manages the overall lifecycle of Operators and their resources, by doing things like resolving dependencies on other Operators, triggering updates to both an Operator and the application it manages, or granting a team access to an Operator for their slice of the cluster.

> Simple, stateless applications can use the Lifecycle Management features of the Operator Framework—without writing any code—by using a generic Operator (for example, the Helm Operator). However, complex and stateful applications are where an Operator can be especially useful. The managed-service capabilities that are encoded into the Operator code can provide an advanced user experience, automating such features as updates, backups and scaling.

Reviewer 2: ```suggestion
Simple, stateless applications can use the Lifecycle Management features of the Operator Framework—without writing any code—by using a generic Operator (for example, the Helm Operator). However, complex and stateful applications are where an Operator can be especially useful. The managed-service capabilities that are encoded into the Operator code can provide an advanced user experience, automating features such as updates, backups and scaling.
```

> Simple, stateless applications can use the Lifecycle Management features of the Operator Framework—without writing any code—by using a generic Operator (for example, the Helm Operator). However, complex and stateful applications are where an Operator can be especially useful. The managed-service capabilities that are encoded into the Operator code can provide an advanced user experience, automating such features as updates, backups and scaling.

Reviewer 2: ```suggestion
Simple, stateless applications can use the Lifecycle Management features of the Operator Framework — without writing any code — by using a generic Operator (for example, the Helm Operator). However, complex and stateful applications are where an Operator can be especially useful. The managed-service capabilities that are encoded into the Operator code can provide an advanced user experience, automating such features as updates, backups, and scaling.
```

Reviewer 2: Hmm, nevermind this looked strange on this first read, but this renders fine when looking at the markdown preview tab of these changes. Disregard.

> 
> 

Reviewer 2: ```suggestion

```

---
title: "Operators"
linkTitle: "Operators"
weight: 3
date: 2021-03-15
description: >
    What are Operators
---

## What is an Operator?

> Operators make it easy to manage complex stateful applications on top of Kubernetes. By [Extending the Kubernetes API](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) with CustomResourceDefinitions it is possible to create what we call Operator Projects where you will be able to create and define your own API's to represent your solutions on the cluster and manage them by looking for the desired state. For a better understanding, see the [Operator Pattern][operator-pattern].

Reviewer 4: Why not use the definitions from the introduction blog post? Those are much more refined and pretty much approved for enterprise customers.

> Operators make it easy to manage complex stateful applications on top of Kubernetes. By [Extending the Kubernetes API](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) with CustomResourceDefinitions it is possible to create what we call Operator Projects where you will be able to create and define your own API's to represent your solutions on the cluster and managed them by looking for the desired state. For a better understanding, see the [Operator Pattern][operator-pattern].

Reviewer 5: Nit: 
```suggestion
Operators make it easy to manage complex stateful applications on top of Kubernetes. By [Extending the Kubernetes API](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) with CustomResourceDefinitions it is possible to create what we call Operator Projects where you will be able to create and define your own API's to represent your solutions on the cluster and manage them by looking for the desired state. For a better understanding, see the [Operator Pattern][operator-pattern].
```

## What is the difference between a Controller and an Operator

> See that following the [Operator's pattern][operator-pattern]; you will create [Controllers](https://kubernetes.io/docs/concepts/architecture/controller/) which provides a reconcile function responsible for synchronizing the resources until a desired state on the cluster. For a more practical explanation, see also [What’s in a controller?](https://book.kubebuilder.io/cronjob-tutorial/controller-overview.html#whats-in-a-controller).

Reviewer 5: This doesn't exactly state the difference between a controller and an operator and instead points towards doc to explain what a controller is and how it can be built. Could you refactor this part or maybe change the question for this answer?
If we are having this question, we could provide a short explanation that operators have application specific controllers, and native Kubernetes controllers are just reconcilers. (Something on similar lines).

Reviewer 4: +1

Reviewer 4: +1

## What is Operator Lifecycle Manager?

> Operator Lifecycle Manager (OLM) helps users install, update, and manage the lifecycle of all Operators and their associated services running across their clusters. This project is a component of the [Operator Framework][operator-framework], an open-source toolkit to manage Kubernetes native applications, called Operators, in an effective, automated, and scalable way. Read more in the introduction [blog post](https://www.openshift.com/blog/introducing-the-operator-framework?extIdCarryOver=true&sc_cid=701f2000001Css5AAC).

Reviewer 4: There's a definition here that goes into much more detail: https://github.com/operator-framework/operator-lifecycle-manager#overview

> ## How does OLM enable cluster admins and developers?

Reviewer 4: This question can be bundled up with the previous question when we use the definition from the olm repo linked above.

The OLM is enabled by default in OpenShift Container Platform 4.x, which aids cluster administrators in installing, upgrading, and granting access to Operators running on their cluster. The OpenShift Container Platform (OCP) web console provides management screens for cluster administrators to install Operators, as well as grant specific projects access to use the catalog of Operators available on the cluster.

For developers, a self-service experience allows provisioning and configuring instances of databases, monitoring, and big data services without having to be subject matter experts, because the Operator has that knowledge baked into it.

> ## Build with the Operator SDK

Reviewer 4: From the blog post mentioned this is doc: How do I start writing an Operator? 
seems like a more appropriate question to start off this section?

[Operator SDK][osdk] is also a component of the [Operator Framework][operator-framework]. Writing an operator today can be difficult because of challenges such as using low-level APIs, writing boilerplate, and a lack of modularity which leads to duplication.
The [Operator SDK][osdk] is a framework that uses the [controller-runtime][controller-runtime] library to make writing operators easier by providing:

- High-level APIs and abstractions to write the operational logic more intuitively
- Tools for scaffolding and code generation to bootstrap a new project fast
- Extensions to cover common operator use cases

Note that [Operator SDK][osdk] provides helpers and features to help you [integrate your project with OLM][olm-integration].

> ## Package with the Operator Lifecycle Manager

Reviewer 4: This should probably link to the Getting started page of this site. 
Packaging an operator for olm is quite an involved process that I'm not sure can be summarized in a paragraph. (this entire site is about packaging an operator for OLM)
